dali/operators/image/color/debayer_gpu.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include "dali/core/span.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/imgproc/color_manipulation/debayer/debayer.h"
+#include "dali/kernels/imgproc/color_manipulation/debayer/debayer_npp.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/operators/image/color/debayer.h"
+
+namespace dali {
+
+namespace debayer {
+
+template <typename InOutT>
+class DebayerImplGPU : public DebayerImplBase<GPUBackend> {
+ public:
+  using Kernel = kernels::debayer::NppDebayerKernel<InOutT>;
+  explicit DebayerImplGPU(const OpSpec &spec, const std::vector<debayer::DALIBayerPattern> &pattern)
+      : pattern_{pattern} {
+    kmgr_.Resize<Kernel>(1, spec.template GetArgument<int>("device_id"));
+  }
+
+  void RunImpl(Workspace &ws) override {
+    auto &output = ws.Output<GPUBackend>(0);
+    output.SetLayout("HWC");
+    ctx_.gpu.stream = ws.stream();
+
+    auto out_view = view<InOutT, 3>(output);
+    auto in_view = GetInView(ws);
+    kmgr_.Run<Kernel>(0, ctx_, out_view, in_view, make_span(pattern_));
+  }
+
+ protected:
+  TensorListView<StorageGPU, const InOutT, 2> GetInView(Workspace &ws) {
+    const auto &input = ws.Input<GPUBackend>(0);
+    const auto &in_shape = input.shape();
+    assert(in_shape.sample_dim() == 2 ||
+           in_shape.sample_dim() == 3);  // by Debayer op's base class setupimpl
+    if (in_shape.sample_dim() == 2) {
+      return view<const InOutT, 2>(input);
+    }
+    assert(([&]() {
+      for (int sample_idx = 0; sample_idx < in_shape.num_samples(); sample_idx++) {
```

Comment:
```C++
assert(std::all_of(in_shape.begin(), in_shape.end(), [](auto &shape) { return shape[2] == 1; }));
```

nitpick

Reply:
Unfortunately, AFAIK, there is no begin/end in neither TL nor TLS class.